### PR TITLE
feat(aws/cloudformation): Add cloudformation controller

### PR DESCRIPTION
**Pull request by xavileon**
This API will be used by Orca to query the status of a cloudformation
deploy operation asynchronously. It uses the cloud formation provider
to access the cached value by the cloudformation caching agent.

This PR partly addresses spinnaker/spinnaker#736 and depends on https://github.com/spinnaker/clouddriver/pull/3278

Signed-off-by: Xavi León <xavi.leon@schibsted.com>

### CODE REVIEW

File: clouddriver-web/src/main/groovy/com/netflix/spinnaker/clouddriver/controllers/CloudFormationController.java [hunk 1]
```diff
@@ -14,42 +14,51 @@
...
   CloudFormation get(@RequestParam String stackId) {
-    log.debug("Cloud formation get stack with id $stackId")
-    try {
-      cloudFormationProviders.collect {
-        it.get(stackId)
-      }.first()
-    } catch (NoSuchElementException e) {
-      throw new ResourceNotFoundException("Cloud Formation stackId $stackId not found.")
-    }
+    log.debug("Cloud formation get stack with id $stackId");
```
Comment (jervi) [inline clouddriver-web/src/main/groovy/com/netflix/spinnaker/clouddriver/controllers/CloudFormationController.java]: log.debug("Cloud formation get stack with id {}", stackId);

File: clouddriver-web/src/main/groovy/com/netflix/spinnaker/clouddriver/controllers/CloudFormationController.java [hunk 2]
```diff
@@ -14,42 +14,51 @@
...
 
   @RequestMapping(method = RequestMethod.GET, value = "/list/{accountId}")
   List<CloudFormation> list(@PathVariable String accountId,
-                            @RequestParam(required = false, defaultValue = '*') String region) {
-    log.debug("Cloud formation list stacks for account $accountId")
-    cloudFormationProviders.collect {
-      it.list(accountId, region)
-    }.flatten()
+                            @RequestParam(required = false, defaultValue = "*") String region) {
+    log.debug("Cloud formation list stacks for account $accountId");
```
Comment (jervi) [inline clouddriver-web/src/main/groovy/com/netflix/spinnaker/clouddriver/controllers/CloudFormationController.java]: log.debug("Cloud formation list stacks for account {}", accountId);

File: clouddriver-web/src/main/groovy/com/netflix/spinnaker/clouddriver/controllers/CloudFormationController.java [hunk 3]
```diff
@@ -14,42 +14,51 @@
...
-      throw new ResourceNotFoundException("Cloud Formation stackId $stackId not found.")
-    }
+    log.debug("Cloud formation get stack with id $stackId");
+    return cloudFormationProviders
+      .stream()
+      .map(p -> p.get(stackId))
+      .filter(Optional::isPresent)
+      .map(Optional<CloudFormation>::get)
+      .findFirst()
+      .<ResourceNotFoundException>orElseThrow(() -> {
```
Comment (jervi) [inline clouddriver-web/src/main/groovy/com/netflix/spinnaker/clouddriver/controllers/CloudFormationController.java]: I don't think you need the generic stuff here, just do `.orElseThrow(() -> {`

File: clouddriver-web/src/main/groovy/com/netflix/spinnaker/clouddriver/controllers/CloudFormationController.java [hunk 4]
```diff
@@ -14,42 +14,51 @@
...
-import groovy.util.logging.Slf4j
-import org.springframework.beans.factory.annotation.Autowired
-import org.springframework.data.rest.webmvc.ResourceNotFoundException
-import org.springframework.web.bind.annotation.*
+import com.netflix.spinnaker.clouddriver.model.CloudFormation;
+import com.netflix.spinnaker.clouddriver.model.CloudFormationProvider;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.rest.webmvc.ResourceNotFoundException;
+import org.springframework.web.bind.annotation.*;
```
Comment (jervi) [inline clouddriver-web/src/main/groovy/com/netflix/spinnaker/clouddriver/controllers/CloudFormationController.java]: I'd split this into separate imports

**Comment by xavileon** [comment]
thanks @jervi addressed your comments in https://github.com/spinnaker/clouddriver/pull/3279/commits/85db3826d96f169122c8bb9361fc7690b2b74c28

File: clouddriver-web/src/main/groovy/com/netflix/spinnaker/clouddriver/controllers/CloudFormationController.java (hunk 3 above)
Comment (robzienert) [reply clouddriver-web/src/main/groovy/com/netflix/spinnaker/clouddriver/controllers/CloudFormationController.java]: I think you can just do this? `orElseThrow(() -> new ResourceNotFoundException("..."));`

Comment (xavileon) [reply clouddriver-web/src/main/groovy/com/netflix/spinnaker/clouddriver/controllers/CloudFormationController.java]: 👍 Done in https://github.com/spinnaker/clouddriver/pull/3279/commits/1e91224e4c0d384adb0ff8ddeb192a120b7411c0

**Pull request merged**